tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_runner.cc

```diff
@@ -0,0 +1,272 @@
+    const HloInstruction* instr) {
+  absl::string_view target = instr->custom_call_target();
+  if (target == kCudnnBatchNormForwardInferenceCallTarget) {
+    return CudnnBatchNormKind::kCudnnBatchNormForwardInference;
+  }
+  if (target == kCudnnBatchNormForwardTrainingCallTarget) {
+    return CudnnBatchNormKind::kCudnnBatchNormForwardTraining;
+  }
+  if (target == kCudnnBatchNormBackwardCallTarget) {
+    return CudnnBatchNormKind::kCudnnBatchNormBackward;
+  }
+  return InternalError("Unexpected call target: %s", target);
+}
+
+void AssignNonOptionalParams(CudnnBatchNormParams& params,
+                             const se::DeviceMemoryBase& operand,
+                             const se::DeviceMemory<float>& scale,
+                             float epsilon) {
+  params.operand = operand;
+  params.scale = scale;
+  params.epsilon = epsilon;
+}
+
+static std::pair<se::dnn::BatchDescriptor /*input_desc*/,
+                 se::dnn::BatchDescriptor /*scale_offset_desc*/>
+MakeBatchNormDescriptors(const Shape& shape, int64 feature_index) {
+  std::vector<int64> logical_to_physical =
+      LayoutUtil::MakeLogicalToPhysical(shape.layout());
+
+  auto physical_dim_size = [&](int64 physical_dim) {
+    return shape.dimensions(LayoutUtil::Major(shape.layout(), physical_dim));
+  };
+
+  // Batchnorm only cares about the location of the depth (aka "feature") dim.
+  // The other dims are all treated the same.  Thus we can use the kBatchDepthYX
+  // cudnn layout for any XLA shape+layout, even XLA shapes that don't have
+  // exactly 4 dimensions: We put everything that comes before the feature dim
+  // into "batch", and everything that comes after the feature dim into "Y".
+  int64 batch_size = 1;
+  int64 y_size = 1;
+  int64 physical_dim;
+  for (physical_dim = 0; physical_dim != logical_to_physical[feature_index];
+       ++physical_dim) {
+    CHECK_LT(physical_dim, shape.dimensions_size());
+    batch_size *= physical_dim_size(physical_dim);
+  }
+  ++physical_dim;  // Skip the feature dimension.
+  for (; physical_dim < shape.dimensions_size(); ++physical_dim) {
+    y_size *= physical_dim_size(physical_dim);
+  }
+
+  se::dnn::BatchDescriptor input_desc;
+  input_desc.set_layout(se::dnn::DataLayout::kBatchDepthYX)
+      .set_count(batch_size)
+      .set_feature_map_count(shape.dimensions(feature_index))
+      .set_height(y_size)
+      .set_width(1);
+
+  se::dnn::BatchDescriptor scale_offset_desc;
+  scale_offset_desc.set_layout(se::dnn::DataLayout::kBatchDepthYX)
+      .set_feature_map_count(input_desc.feature_map_count())
+      .set_height(1)
+      .set_width(1)
+      .set_count(1);
+
+  return std::make_pair(input_desc, scale_offset_desc);
+}
+
+template <typename ElemType>
+Status RunCudnnBatchNormImplInternal(CudnnBatchNormParams& params,
+                                     se::Stream* stream) {
+  se::DeviceMemory<float> null_device_ptr(nullptr);
+  switch (params.kind) {
```

Comment:
Following up after reading the header, I'm not sure if having a common dispatcher like this is useful (it actually obscures the flow a bit).

How about just having three templated functions: `RunCudnnBatchNorm{ForwardInference|ForwardTraining|Backward}ImplInternal` and having the exposed API call into the right one with a switch on the floating point type?  That will also let us use stricter types instead of a single common `CudnnBatchNormParams` as mentioned in the header.

Reply:
This makes sense. I thought about this initially but didn't go through with this to make this in-line with `cudnn_conv_runner` (now `gpu_conv_runner`). I've made the changes. Let me knw if there are any further modifications required here.